lightning/src/util/indexed_map.rs

```diff
@@ -0,0 +1,156 @@
+
+use alloc::collections::{BTreeMap, btree_map};
+use core::cmp::Ord;
+use core::ops::RangeBounds;
+
+/// A map which can be iterated in a deterministic order.
+///
+/// This would traditionally be accomplished by simply using a `BTreeMap`, however a B-Trees
+/// generally have very slow lookups. Because we use a nodes+channels map while finding routes
+/// across the network graph, our network graph backing map must be as performant as possible.
+/// However, because peers expect to sync the network graph from us (and we need to support that
+/// without holding a lock on the graph for the duration of the sync or dumping the entire graph
+/// into our outbound message queue), we need an iterable map with a consistent iteration order we
+/// can jump to a starting point on.
+///
+/// Thus, we have a custom data structure here - its API mimics that of Rust's `BTreeMap`, but is
+/// actually backed by a `HashMap`, with some additional tracking to ensure we can iterate over
+/// keys in the order defined by [`Ord`].
+#[derive(Clone, PartialEq, Eq)]
+pub struct IndexedMap<K: Ord, V> {
+	map: BTreeMap<K, V>,
+}
+
+impl<K: Ord, V> IndexedMap<K, V> {
+	/// Constructs a new, empty map
+	pub fn new() -> Self {
+		Self {
+			map: BTreeMap::new(),
+		}
+	}
+
+	#[inline(always)]
+	/// Fetches the element with the given `key`, if one exists.
+	pub fn get(&self, key: &K) -> Option<&V> {
+		self.map.get(key)
+	}
+
+	/// Fetches a mutable reference to the element with the given `key`, if one exists.
+	pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
+		self.map.get_mut(key)
+	}
+
+	#[inline]
+	/// Returns true if an element with the given `key` exists in the map.
+	pub fn contains_key(&self, key: &K) -> bool {
+		self.map.contains_key(key)
+	}
+
+	/// Removes the element with the given `key`, returning it, if one exists.
+	pub fn remove(&mut self, key: &K) -> Option<V> {
+		self.map.remove(key)
+	}
+
+	/// Inserts the given `key`/`value` pair into the map, returning the element that was
+	/// previously stored at the given `key`, if one exists.
+	pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+		self.map.insert(key, value)
+	}
+
+	/// Returns an [`Entry`] for the given `key` in the map, allowing access to the value.
+	pub fn entry(&mut self, key: K) -> Entry<'_, K, V> {
+		match self.map.entry(key) {
+			btree_map::Entry::Vacant(entry) => {
+				Entry::Vacant(VacantEntry {
+					underlying_entry: entry
+				})
+			},
+			btree_map::Entry::Occupied(entry) => {
+				Entry::Occupied(OccupiedEntry {
+					underlying_entry: entry
+				})
+			}
+		}
+	}
+
+	/// Returns an iterator which iterates over the keys in the map, in a random order.
+	pub fn unordered_keys(&self) -> impl Iterator<Item = &K> {
+		self.map.keys()
+	}
+
+	/// Returns an iterator which iterates over the `key`/`value` pairs in a random order.
+	pub fn unordered_iter(&self) -> impl Iterator<Item = (&K, &V)> {
```

Comment:
what's the benefit of a random order? I thought the entire point of this data structure was that the iteration order be deterministic?

Reply:
But won't it technically always return an ordered version? Considering that, at least in this commit, the underlying structure is a BTreeMap?

Reply:
And that's… a good thing?